.circleci/config.yml

```diff
@@ -22,21 +31,217 @@ anchors:
+            # If that fails, then check for the latest from master branch.
+            # If that fails, then check for the latest from any branch
+            - << parameters.version >>-yarn-ui-cache-{{ checksum "~/workbench/ui/yarn.lock" }}-{{ .Branch }}
+            - << parameters.version >>-yarn-ui-cache-{{ checksum "~/workbench/ui/yarn.lock" }}-master
+            - << parameters.version >>-yarn-ui-cache-{{ checksum "~/workbench/ui/yarn.lock" }}
+      - run:
+          name: "workbench/ui: yarn install dependencies"
+          working_directory: ~/workbench/ui
+          command: yarn install --verbose --frozen-lockfile --non-interactive
+      - save_cache:
+          paths:
+            - ~/.cache/yarn
+            - ui/node_modules
+          key: << parameters.version >>-yarn-ui-cache-{{ checksum "~/workbench/ui/yarn.lock" }}-{{ .Branch }}
+
+  e2e_dependencies:
+    description: "workbench/e2e: yarn install, save and restore cache"
+    parameters:
+      version:
+        type: string
+        default: << pipeline.parameters.cache_key_version >>
+    steps:
       - restore_cache:
           keys:
-          - api-cache-{{ checksum "~/workbench/api/build.gradle" }}
-          - api-cache-
+            - << parameters.version >>-yarn-e2e-cache-{{ checksum "~/workbench/e2e/yarn.lock" }}-{{ .Branch }}
+            - << parameters.version >>-yarn-e2e-cache-{{ checksum "~/workbench/e2e/yarn.lock" }}-master
+            - << parameters.version >>-yarn-e2e-cache-{{ checksum "~/workbench/e2e/yarn.lock" }}
+      - run:
+          name: "workbench/e2e: yarn install dependencies"
+          working_directory: ~/workbench/e2e
+          command: yarn install --frozen-lockfile --non-interactive
+      - save_cache:
+          paths:
+            - ~/.cache/yarn
+            - e2e/node_modules
+          key: << parameters.version >>-yarn-e2e-cache-{{ checksum "~/workbench/e2e/yarn.lock" }}-{{ .Branch }}
+
+  api_cache:
+    description: "workbench/api: save and restore gradle cache"
+    parameters:
+      save:
+        type: boolean
+        default: false
+      restore:
+        type: boolean
+        default: false
+    steps:
+      - when:
+          condition: << parameters.save >>
+          steps:
+            - save_cache:
+                paths:
+                  - ~/.gradle/caches
+                  - ~/.gradle/wrapper
+                  - ~/.m2
+                  - ~/workbench/api/build/exploded-api/WEB-INF/lib/
+                key: v1-api-cache-{{ checksum "~/workbench/api/build.gradle" }}-{{ .Branch }}
+      - when:
+          condition: << parameters.restore >>
+          steps:
+            - restore_cache:
+                keys:
+                  - v1-api-cache-{{ checksum "~/workbench/api/build.gradle" }}-{{ .Branch }}
+                  - v1-api-cache-{{ checksum "~/workbench/api/build.gradle" }}-master
+                  - v1-api-cache-{{ checksum "~/workbench/api/build.gradle" }}
+
+  deploy_appengine:
```

Comment:
I'm not able to test any `deploy to app engine` jobs since they're only for master and release branches.